README.md
# kinesis-write-stream

Kinesis writable stream that buffers up records

[![build status](https://travis-ci.org/voldern/kinesis-write-stream.svg)](https://travis-ci.org/voldern/kinesis-write-stream)
[![modules status](https://david-dm.org/voldern/kinesis-write-stream.svg)](https://david-dm.org/voldern/kinesis-write-stream)

[![npm badge](https://nodei.co/npm/kinesis-write-stream.png?downloads=true)](https://nodei.co/npm/kinesis-write-stream)

# Usage

Records written to the stream will buffer up until `highWaterMark` has
been reached, or the stream is closed, before writing to Kinesis using
`putRecords`.

## Failed items

Failed items will be retried up until `options.maxRetries` has been
reached. The initial timeout before the retry is set in
`options.retryTimeout` and it increases by the fibonnaci sequence.

## Partition key

The partition key is by default a random number, but this can be
adjusted by overriding `streamInstance.getPartitionKey(record)`.

## Flushing

Its also possible to send in the option `flushTimeout` to indicate
that the items currently in the buffer should be flushed after the
given amount of milliseconds if the `highWaterMark` haven't been
reached.

## Logging

A [bunyan](https://www.npmjs.com/package/bunyan),
[winston](https://www.npmjs.com/package/winston) or similar logger
instance that have methods like `debug`, `error` and `info` may be
sent in as `options.logger` to the constructor.

# Example

```javascript
var KinesisWritable = require('kinesis-write-stream');

var stream = new KinesisWritable(kinesisClient, 'streamName', {
  highWaterMark: 16,
  maxRetries: 3,
  retryTimeout: 100
});

inputStream.pipe(stream);
```

# API

See [api.md](api.md).

# License

MIT


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [1.1.1] - 2018-01-18
### Fixed
- Fixed bug that would cause loss of events written between a write to Kinesis 
  started and finished. Fixed by @chrissharkey.

## [1.1.0] - 2017-05-15
### Added
- Added flushTimeout for periodically flushing records. 
  Thanks to @indieisaconcept

## [1.0.2] - 2017-03-25
### Fixed
- Pass record to getPartitionKey as stated in
  documentation. Contribution by @indieisaconcept.

## [1.0.1] - 2015-11-04
### Fixed
- Call `warn` method instead of `warning` on log object if available

## [1.0.0] - 2015-11-04
### Fixed
- Do not write to Kinesis when queue is empty

## [0.0.1] - 2015-10-28

Initial release


test/kinesis-write-stream.js
'use strict';

var chai = require('chai'),
    sinon = require('sinon'),
    sinonChai = require('sinon-chai'),
    streamArray = require('stream-array'),
    _ = require('lodash'),
    KinesisWritable = require('../'),
    recordsFixture = require('./fixture/records'),
    successResponseFixture = require('./fixture/success-response'),
    errorResponseFixture = require('./fixture/failed-response'),
    writeFixture = require('./fixture/write-fixture');

chai.use(sinonChai);

var expect = chai.expect;

describe('KinesisWritable', function() {
    beforeEach(function() {
        this.sinon = sinon.sandbox.create();

        this.client = {
            putRecords: sinon.stub()
        };

        this.stream = new KinesisWritable(this.client, 'streamName', {
            highWaterMark: 6
        });
    });

    afterEach(function() {
        this.sinon.restore();
    });

    describe('constructor', function() {
        it('should throw error on missing client', function() {
            expect(function() {
                new KinesisWritable();
            }).to.Throw(Error, 'client is required');
        });

        it('should throw error on missing streamName', function() {
            expect(function() {
                new KinesisWritable({});
            }).to.Throw(Error, 'streamName is required');
        });

        it('should throw error on highWaterMark above 500', function() {
            expect(function() {
                new KinesisWritable({}, 'test', { highWaterMark: 501 });
            }).to.Throw(Error, 'Max highWaterMark is 500');
        });
    });

    describe('getPartitionKey', function() {
        it('should return a random partition key padded to 4 digits', function() {
            var kinesis = new KinesisWritable({}, 'foo');

            this.sinon.stub(_, 'random').returns(10);

            expect(kinesis.getPartitionKey()).to.eq('0010');

            _.random.returns(1000);

            expect(kinesis.getPartitionKey()).to.eq('1000');
        });

        it('should be called with the current record being added', function(done) {
            this.client.putRecords.yields(null, successResponseFixture);
            this.sinon.stub(this.stream, 'getPartitionKey').returns('1234');

            this.stream.on('finish', function() {
                expect(this.stream.getPartitionKey).to.have.been.calledWith(recordsFixture[0]);
                done();
            }.bind(this));

            streamArray([recordsFixture[0]])
                .pipe(this.stream);
        });

        it('should use custom getPartitionKey if defined', function(done) {
            this.client.putRecords.yields(null, successResponseFixture);

            this.stream.getPartitionKey = function() {
                return 'custom-partition';
            };

            this.sinon.spy(this.stream, 'getPartitionKey');

            this.stream.on('finish', function() {
                expect(this.stream.getPartitionKey).to.have.returned('custom-partition');
                done();
            }.bind(this));

            streamArray(recordsFixture)
                .pipe(this.stream);
        });
    });

    describe('_write', function() {
        it('should write to kinesis when stream is closed', function(done) {
            this.client.putRecords.yields(null, successResponseFixture);
            this.sinon.stub(this.stream, 'getPartitionKey').returns('1234');

            this.stream.on('finish', function() {
                expect(this.client.putRecords).to.have.been.calledOnce;

                expect(this.client.putRecords).to.have.been.calledWith({
                    Records: writeFixture,
                    StreamName: 'streamName'
                });

                done();
            }.bind(this));

            streamArray(recordsFixture)
                .pipe(this.stream);
        });

        it('should do nothing if there is nothing in the queue when the stream is closed', function(done) {
            this.client.putRecords.yields(null, successResponseFixture);

            this.stream.on('finish', function() {
                expect(this.client.putRecords).to.have.been.calledOnce;

                done();
            }.bind(this));

            for (var i = 0; i < 6; i++) {
                this.stream.write(recordsFixture);
            }

            this.stream.end();
        });

        it('should buffer records up to highWaterMark', function(done) {
            this.client.putRecords.yields(null, successResponseFixture);

            for (var i = 0; i < 4; i++) {
                this.stream.write(recordsFixture[0]);
            }

            this.stream.write(recordsFixture[0], function() {
                expect(this.client.putRecords).to.not.have.been.called;

                this.stream.write(recordsFixture[0], function() {
                    expect(this.client.putRecords).to.have.been.calledOnce;

                    done();
                }.bind(this));
            }.bind(this));
        });

        it('should emit error on Kinesis error', function(done) {
            this.client.putRecords.yields('Fail');

            this.stream.on('error', function(err) {
                expect(err).to.eq('Fail');

                done();
            });

            this.stream.end({ foo: 'bar' });
        });

        it('should emit error on records errors', function(done) {
            this.client.putRecords.yields(null, errorResponseFixture);

            this.stream.on('error', function(err) {
                expect(err).to.be.ok;

                done();
            });

            this.stream.end({ foo: 'bar' });
        });

        it('should retry failed records', function(done) {
            this.sinon.stub(this.stream, 'getPartitionKey').returns('1234');

            this.client.putRecords.yields(null, errorResponseFixture);
            this.client.putRecords.onCall(2).yields(null, successResponseFixture);

            this.stream.on('finish', function() {
                expect(this.client.putRecords).to.have.been.calledThrice;

                expect(this.client.putRecords.secondCall).to.have.been.calledWith({
                    Records: [
                        writeFixture[1],
                        writeFixture[3]
                    ],
                    StreamName: 'streamName'
                });

                done();
            }.bind(this));

            streamArray(recordsFixture).pipe(this.stream);
        });
    });

    describe('flush timeout', function() {
        var originalSetTimeout = setTimeout;

        beforeEach(function() {
            this.sinon = sinon.sandbox.create();

            this.client = {
                putRecords: sinon.stub()
            };

            this.stream = new KinesisWritable(this.client, 'streamName', {
                highWaterMark: 6,
                flushTimeout: 1000
            });

            this.client.putRecords.yields(null, successResponseFixture);
            this.clock = sinon.useFakeTimers();
        });

        it('should flush queue if there is something in the queue after timeout', function() {
            var self = this;

            for (var i = 0; i < 6; i++) {
                this.stream.write(recordsFixture[0]);
            }

            this.clock.tick(0);

            expect(self.client.putRecords).to.have.callCount(1);

            self.stream.write(recordsFixture[0]);
            self.clock.tick(1001);

            expect(self.client.putRecords).to.have.callCount(2);
        });

        it('should clear flush timer when stream has ended', function(done) {
            var self = this;

            this.sinon.spy(this.stream, '_flush');
            this.stream.write(recordsFixture[0]);
            self.clock.tick(0);

            this.stream.end(function() {
                self.clock.tick(0);
                expect(this.stream._flush).to.have.callCount(1);

                this.clock.tick(2001);

                expect(this.stream._flush).to.have.callCount(1);

                done();
            }.bind(this));

            originalSetTimeout(function() {
                self.clock.tick(0);
            });
        });
    });
});


test/fixture/write-fixture.js
'use strict';

var records = require('./records.json');

records = records.map(function(record) {
    return {
        Data: JSON.stringify(record),
        PartitionKey: '1234'
    };
});

module.exports = records;


index.js
'use strict';

var util = require('util'),
    assert = require('assert'),
    retryFn = require('retry-fn'),
    FlushWritable = require('flushwritable'),
    _ = require('lodash');

util.inherits(KinesisWritable, FlushWritable);

/**
 * A simple stream wrapper around Kinesis putRecords
 *
 * @param {AWS.Kinesis} client AWS Kinesis client
 * @param {string} streamName Kinesis stream name
 * @param {Object} options Options
 * @param {number} [options.highWaterMark=16] Number of items to buffer before writing.
 * Also the size of the underlying stream buffer.
 * @param {Number} [options.maxRetries=3] How many times to retry failed data before
 * rasing an error
 * @param {Number} [options.retryTimeout=100] How long to wait initially before
 * retrying failed records
 * @param {number} [options.flushTimeout=null] Number of ms to flush records after, if highWaterMark hasn't been reached
 * @param {Object} [options.logger] Instance of a logger like bunyan or winston
 */
function KinesisWritable(client, streamName, options) {
    assert(client, 'client is required');
    assert(streamName, 'streamName is required');

    options = options || {};
    options.objectMode = true;

    FlushWritable.call(this, options);

    this.client = client;
    this.streamName = streamName;
    this.logger = options.logger || null;
    this.highWaterMark = options.highWaterMark || 16;
    this.flushTimeout = options.flushTimeout || null;
    this.maxRetries = options.maxRetries || 3;

    // The maximum number of items that can be written to Kinesis in
    // one request is 500
    assert(this.highWaterMark <= 500, 'Max highWaterMark is 500');

    this.queue = [];
}

/* eslint-disable no-unused-vars */

/**
 * Get partition key for given record
 *
 * @param {Object} [record] Record
 * @return {string}
 */
KinesisWritable.prototype.getPartitionKey = function getPartitionKey(record) {
    return _.padLeft(_.random(0, 1000), 4, '0');
};

/* eslint-enable no-unused-vars */

/**
 * Transform record into one accepted by Kinesis
 *
 * @private
 * @param {Object} record
 * @return {Object}
 */
KinesisWritable.prototype.transformRecord = function transformRecord(record) {
    return {
        Data: JSON.stringify(record),
        PartitionKey: this.getPartitionKey(record)
    };
};

/**
 * Write records in queue to Kinesis
 *
 * @private
 * @param {Function} callback
 */
KinesisWritable.prototype.writeRecords = function writeRecords(callback) {
    if (this.logger) {
        this.logger.debug('Writing %d records to Kinesis', this.queue.length);
    }

    var recordsToWrite = this.queue.splice(0);
    var records = recordsToWrite.map(this.transformRecord.bind(this));

    this.client.putRecords({
        Records: records,
        StreamName: this.streamName
    }, function(err, response) {
        if (err) {
            return callback(err);
        }

        if (this.logger) {
            this.logger.info('Wrote %d records to Kinesis',
                records.length - response.FailedRecordCount);
        }

        if (response.FailedRecordCount !== 0) {
            if (this.logger) {
                this.logger.warn('Failed writing %d records to Kinesis',
                    response.FailedRecordCount);
            }

            var failedRecords = [];

            response.Records.forEach(function(record, index) {
                if (record.ErrorCode) {
                    if (this.logger) {
                        this.logger.warn('Failed record with message: %s', record.ErrorMessage);
                    }

                    failedRecords.push(recordsToWrite[index]);
                }
            }.bind(this));

            this.queue = this.queue.concat(failedRecords);

            return callback(new Error('Failed to write ' + failedRecords.length + ' records'));
        }

        callback();
    }.bind(this));
};

/**
 * Flush method needed by the underlying stream implementation
 *
 * @private
 * @param {Function} callback
 * @return {undefined}
 */
KinesisWritable.prototype._flush = function _flush(callback) {
    clearTimeout(this.flushTimeoutId);

    if (this.queue.length === 0) {
        return callback();
    }

    var retry = retryFn.bind(null, {
        retries: this.maxRetries,
        timeout: retryFn.fib(this.retryTimeout)
    });

    retry(this.writeRecords.bind(this), callback);
};

/**
 * Write method needed by the underlying stream implementation
 *
 * @private
 * @param {Object} record
 * @param {string} enc
 * @param {Function} callback
 * @returns {undefined}
 */
KinesisWritable.prototype._write = function _write(record, enc, callback) {
    if (this.logger) {
        this.logger.debug('Adding to Kinesis queue', { record: record });
    }
    this.queue.push(record);

    if (this.queue.length >= this.highWaterMark) {
        return this._flush(callback);
    }

    if (this.queue.length >= this.highWaterMark) {
        return this._flush(callback);
    } else if (this.flushTimeout) {
        clearTimeout(this.flushTimeoutId);

        this.flushTimeoutId = setTimeout(function() {
            this._flush(function(err) {
                if (err) {
                    this.emit('error', err);
                }
            }.bind(this));
        }.bind(this), this.flushTimeout);
    }

    callback();
};

module.exports = KinesisWritable;
